### PR TITLE
Compact a Jena database

### PR DESCRIPTION
**Pull request by frankyhollywood**

### CODE REVIEW

File: projects/saturn/src/main/java/io/fairspace/saturn/rdf/SaturnDatasetFactory.java [hunk 1]
```diff
@@ -15,23 +15,23 @@
...
+     * Currently it adds transaction logging and applies default vocabulary if
+     * needed.
      */
     public static Dataset connect(Config.Jena config, ViewStoreClientFactory viewStoreClientFactory) {
         var restoreNeeded = isRestoreNeeded(config.datasetPath);
 
         // Create a TDB2 dataset graph
-        var dsg = connectCreate(Location.create(config.datasetPath.getAbsolutePath()), config.storeParams, null)
-                .getDatasetGraph();
+        Dataset ds = TDB2Factory.connectDataset(config.datasetPath.getAbsolutePath());
```
Comment (tgreenwood) [inline projects/saturn/src/main/java/io/fairspace/saturn/rdf/SaturnDatasetFactory.java]: store pamas got lost

File: projects/saturn/src/main/java/io/fairspace/saturn/services/maintenance/MaintenanceService.java [hunk 2]
```diff
@@ -70,6 +75,38 @@ public synchronized void startRecreateIndexTask() {
         });
     }
 
+    public void compactRdfStorageTask() {
+        if (!userService.currentUser().isAdmin()) {
+            throw new AccessDeniedException();
+        }
+        if (disabled()) {
```
Comment (tgreenwood) [inline projects/saturn/src/main/java/io/fairspace/saturn/services/maintenance/MaintenanceService.java]: this check is redundant for Jena

File: projects/saturn/src/main/java/io/fairspace/saturn/services/maintenance/MaintenanceService.java [hunk 3]
```diff
@@ -70,6 +75,38 @@ public synchronized void startRecreateIndexTask() {
...
+            log.info("Reindexing is already in progress.");
+            throw new ConflictException(REINDEXING_IS_ALREADY_IN_PROGRESS);
+        }
+
+        threadpool.submit(() -> {
+            log.info("Compacting RDF storage started");
+            try {
+                var ds = unwrap(dataset.asDatasetGraph());
+                if (ds == null) {
+                    log.info("Compacting RDF storage is not supported for this storage type");
```
Comment (tgreenwood) [inline projects/saturn/src/main/java/io/fairspace/saturn/services/maintenance/MaintenanceService.java]: maybe to have it as a warn, not info

File: projects/saturn/src/main/java/io/fairspace/saturn/services/maintenance/MaintenanceService.java [hunk 4]
```diff
@@ -70,6 +75,38 @@ public synchronized void startRecreateIndexTask() {
...
 
+    public void compactRdfStorageTask() {
+        if (!userService.currentUser().isAdmin()) {
+            throw new AccessDeniedException();
+        }
+        if (disabled()) {
+            throw new NotAvailableException(SERVICE_NOT_AVAILABLE);
+        }
+        if (active()) {
+            log.info("Reindexing is already in progress.");
```
Comment (tgreenwood) [inline projects/saturn/src/main/java/io/fairspace/saturn/services/maintenance/MaintenanceService.java]: I would replace 'reindexing' with 'maintenance'

File: projects/saturn/src/main/java/io/fairspace/saturn/rdf/SaturnDatasetFactory.java (hunk 1 above)
Comment (frankyhollywood) [reply projects/saturn/src/main/java/io/fairspace/saturn/rdf/SaturnDatasetFactory.java]: fixed it...

File: projects/saturn/src/main/java/io/fairspace/saturn/services/maintenance/MaintenanceService.java [hunk 5]
```diff
@@ -22,7 +27,7 @@
 @Log4j2
 public class MaintenanceService {
     public static final String SERVICE_NOT_AVAILABLE = "Service not available";
-    public static final String REINDEXING_IS_ALREADY_IN_PROGRESS = "Reindexing is already in progress.";
+    public static final String REINDEXING_IS_ALREADY_IN_PROGRESS = "Maintenance is already in progress.";
```
Comment (tgreenwood) [inline projects/saturn/src/main/java/io/fairspace/saturn/services/maintenance/MaintenanceService.java]: I would have two separate messages to distinct in logs which process is ongoing. Then aprove

File: projects/saturn/src/main/java/io/fairspace/saturn/services/maintenance/MaintenanceService.java [hunk 6]
```diff
@@ -102,6 +102,10 @@ public void compactRdfStorageTask() {
         });
     }
 
+    public boolean isMaintenanceInProgress() {
```
Comment (tgreenwood) [inline projects/saturn/src/main/java/io/fairspace/saturn/services/maintenance/MaintenanceService.java]: looks like not used method
maybe just rename 'active'?

**Review by tgreenwood — APPROVED** [review]

**Pull request merged**